std/range/package.d

```diff
@@ -7118,9 +7124,66 @@ private:
     size_t _chunkSize;
 }
 
+/// ditto
+struct Chunks(Source)
+if (isInputRange!Source && !isForwardRange!Source)
+{
+    import std.typecons : RefCounted;
+
+    static struct Chunk
+    {
+        private RefCounted!Impl impl;
+
+        @property bool empty() { return impl.curSizeLeft == 0 || impl.r.empty; }
+        @property auto front() { return impl.r.front; }
+        void popFront()
+        {
+            assert(impl.curSizeLeft > 0 && !impl.r.empty);
+            impl.curSizeLeft--;
+            impl.r.popFront();
+        }
+    }
+
+    static struct Impl
+    {
+        private Source r;
+        private size_t chunkSize;
+        private size_t curSizeLeft;
+        private bool _empty;
+    }
+
+    private RefCounted!Impl impl;
+    alias impl this;
+
+    private this(Source r, size_t chunkSize)
+    {
+        impl = RefCounted!Impl(r, chunkSize, chunkSize);
+        impl._empty = r.empty;
+    }
+
+    @property bool empty() { return impl._empty; }
+    @property Chunk front() return { return Chunk(impl); }
+
+    void popFront()
+    {
+        while (!impl.r.empty && impl.curSizeLeft > 0)
+        {
+            impl.curSizeLeft--;
+            impl.r.popFront();
+        }
+
+        if (!impl.r.empty)
+            impl.curSizeLeft = impl.chunkSize;
+        else
+            impl._empty = true;
```

Comment:
Looks like `impl._empty` is just a cache for `impl.r.empty`, why not eliminate? Alternatively, put a 0 in `impl.chunkSize` when done.